### PR TITLE
perf: Minor optimizations

### PR DESCRIPTION
**Pull request by hangy**
Some minor performance optimizations I stumbled upon while having a look at #341:

1. `FallbackLanguagesCollection`
   1. Avoid dictionary lookup for the `"default"` fallback case
   2. Avoid `ContainsKey`/`[key]` lookup by shortcutting with `TryGetValue`
2. `JsonConverter`
   1. Replaced `ICollection<string>` with `List<string>` to avoid unnecessary virtual method calls
   2. Use `char` instead of `string`
   3. Optimize some string/collection operations with range operator and `Span<T>`
3. `LocalizationProvider`
   1. Use source-generated compiled regex to detect placeholders
   2. Use a token-based approach to replace placeholders to avoid regex allocations

### CODE REVIEW

File: common/src/DbLocalizationProvider/LocalizationProvider.cs [hunk 1]
```diff
@@ -439,15 +439,15 @@ private static string FormatWithAnonymousObject(string message, object model)
             return string.Format(message, model);
         }
 
-        var placeHolders = Regex.Matches(message, "{.*?}").Select(m => m.Value).ToList();
+        var placeHolders = PlaceHolderRegex().Matches(message).Select(m => m.Value).ToList();
 
         if (!placeHolders.Any())
```
Comment (stefanolsen) [inline common/src/DbLocalizationProvider/LocalizationProvider.cs:444]: @hangy Might as well skip Select here. And test that .Matches(message).Count > 0. Then use .Value further down. Saves a lot of allocations by waiting.

Comment (stefanolsen) [reply common/src/DbLocalizationProvider/LocalizationProvider.cs:444]: Also, if you don't actually need the curly slashes now or later, then make a capture group between the curly slashes, so you only get the content of them.

File: common/src/DbLocalizationProvider/Json/JsonConverter.cs [hunk 2]
```diff
@@ -172,7 +174,7 @@ private static string CamelCase(string that)
     {
         if (that.Length > 1)
         {
-            return that.Substring(0, 1).ToLower() + that.Substring(1);
+            return string.Concat(that[..1].ToLower(), that.AsSpan(1));
```
Comment (stefanolsen) [inline common/src/DbLocalizationProvider/Json/JsonConverter.cs:177]: @hangy I usually make my ToCamelCase method like this:
```
ArgumentNullException.ThrowIfNull(text);

if (text.Length == 0 || char.IsLower(text, 0))
{
    return text;
}

return string.Create(text.Length, text, (chars, state) =>
{
    state.AsSpan().CopyTo(chars);
    chars[0] = char.ToLower(chars[0]);
});
```

The existing implementation does not skip ToLower, if the string is empty or the first character is already lower case.

Comment (hangy) [reply common/src/DbLocalizationProvider/Json/JsonConverter.cs:177]: I think the main optimization of that version might be the use of `string.Create`? Benchmarked it against the existing version, and it wins by a huge margin. Nice suggestion!

<details><summary>Benchmark</summary>

```csharp
using BenchmarkDotNet.Attributes;

namespace CamelCaseBenchmarks;

[MemoryDiagnoser]
public class SomeBenchmark
{
    private const string s_singleLowerChar = "A";

    private const string s_singleUpperChar = "a";

    private const string s_alreadyCamelCase = "someStringIsCamelCaseAndSuperCoolAndDoesNotNeedConversion";

    private const string s_notCamelCase = "SomeStringIsNotCamelCaseAndNeedsConversion";

    [Benchmark]
    public void BenchmarkSomeAlreadyCamelCaseOriginal() => CamelCase(s_alreadyCamelCase);

    [Benchmark]
    public void BenchmarkSomeAlreadyCamelCaseSuggested() => CamelCaseSuggested(s_alreadyCamelCase);

    [Benchmark]
    public void BenchmarkSomeNotAlreadyCamelCaseOriginal() => CamelCase(s_notCamelCase);

    [Benchmark]
    public void BenchmarkSomeNotAlreadyCamelCaseSuggested() => CamelCaseSuggested(s_notCamelCase);

    [Benchmark]
    public void BenchmarkSingleLowerCharOriginal() => CamelCase(s_singleLowerChar);

    [Benchmark]
    public void BenchmarkSingleLowerCharSuggested() => CamelCaseSuggested(s_singleLowerChar);

    [Benchmark]
    public void BenchmarkSingleUpperCharOriginal() => CamelCase(s_singleUpperChar);

    [Benchmark]
    public void BenchmarkSingleUpperCharSuggested() => CamelCaseSuggested(s_singleUpperChar);

    private static string CamelCase(string that)
    {
        if (that.Length > 1)
        {
            return string.Concat(that[..1].ToLower(), that.AsSpan(1));
        }

        return that.ToLower();
    }

    private static string CamelCaseSuggested(string text)
    {
        ArgumentNullException.ThrowIfNull(text);

        if (text.Length == 0 || char.IsLower(text, 0))
        {
            return text;
        }

        return string.Create(text.Length, text, (chars, state) =>
        {
            state.AsSpan().CopyTo(chars);
            chars[0] = char.ToLower(chars[0]);
        });
    }
}

```

| Method                                    | Mean      | Error     | StdDev    | Gen0   | Allocated |
|------------------------------------------ |----------:|----------:|----------:|-------:|----------:|
| BenchmarkSomeAlreadyCamelCaseOriginal     | 16.965 ns | 0.3113 ns | 0.2912 ns | 0.0032 |     160 B |
| BenchmarkSomeAlreadyCamelCaseSuggested    |  1.071 ns | 0.0118 ns | 0.0110 ns |      - |         - |
| BenchmarkSomeNotAlreadyCamelCaseOriginal  | 21.448 ns | 0.1997 ns | 0.1868 ns | 0.0032 |     160 B |
| BenchmarkSomeNotAlreadyCamelCaseSuggested |  8.502 ns | 0.0859 ns | 0.0803 ns | 0.0022 |     112 B |
| BenchmarkSingleLowerCharOriginal          | 10.304 ns | 0.0650 ns | 0.0576 ns | 0.0005 |      24 B |
| BenchmarkSingleLowerCharSuggested         |  6.786 ns | 0.0612 ns | 0.0572 ns | 0.0005 |      24 B |
| BenchmarkSingleUpperCharOriginal          |  4.430 ns | 0.0356 ns | 0.0333 ns |      - |         - |
| BenchmarkSingleUpperCharSuggested         |  1.054 ns | 0.0089 ns | 0.0074 ns |      - |         - |

</detail>

File: common/src/DbLocalizationProvider/LocalizationProvider.cs (hunk 1 above)
Comment (hangy) [reply common/src/DbLocalizationProvider/LocalizationProvider.cs:444]: Not as straight-forward as I may have hoped. Capturing the regex groups probably causes additional allocations. It's faster for no placeholder or repeated placeholders but slower for a simple string with two placeholders.

<details><summary>WIP Benchmarks</summary>

```csharp
using System.Text.RegularExpressions;
using BenchmarkDotNet.Attributes;

namespace CamelCaseBenchmarks;

[MemoryDiagnoser]
public partial class SomeBenchmark
{
    private const string MessageNoPlaceHolder = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.";

    private const string Message = "Hello {Name}, your age is {Age}";

    private const string MessageRepeatedPlaceHolder = "Hello {Name}, your age is {Age} {Name} {Age} {Name} {Age} {Name} {Age} {Name} {Age} {Name} {Age} {Name} {Age} {Name} {Age} {Name} {Age}";

    private static readonly object s_model = new { Name = "John", Age = 30, Address = "Some address", City = "Some city" };

    [Benchmark]
    public string FormatWithAnonymousObjectNoPlaceHolderOrig() => FormatWithAnonymousObjectOrig(MessageNoPlaceHolder, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectOrig() => FormatWithAnonymousObjectOrig(Message, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectRepeatedPlaceHolderOrig() => FormatWithAnonymousObjectOrig(MessageRepeatedPlaceHolder, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectNoPlaceHolderOld() => FormatWithAnonymousObjectOld(MessageNoPlaceHolder, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectOld() => FormatWithAnonymousObjectOld(Message, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectRepeatedPlaceHolderOld() => FormatWithAnonymousObjectOld(MessageRepeatedPlaceHolder, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectNoPlaceHolderNew() => FormatWithAnonymousObjectNew(MessageNoPlaceHolder, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectNew() => FormatWithAnonymousObjectNew(Message, s_model);

    [Benchmark]
    public string FormatWithAnonymousObjectRepeatedPlaceHolderNew() => FormatWithAnonymousObjectNew(MessageRepeatedPlaceHolder, s_model);

    private static string FormatWithAnonymousObjectOrig(string message, object model)
    {
        var type = model.GetType();
        if (type == typeof(string))
        {
            return string.Format(message, model);
        }

        var placeHolders = Regex.Matches(message, "{.*?}").Select(m => m.Value).ToList();

        if (!placeHolders.Any())
        {
            return message;
        }

        var placeholderMap = new Dictionary<string, object>();
        var properties = type.GetProperties();

        foreach (var placeHolder in placeHolders)
        {
            var propertyInfo = properties.FirstOrDefault(p => p.Name == placeHolder.Trim('{', '}'));

            // property found - extract value and add to the map
            var val = propertyInfo?.GetValue(model);
            if (val != null && !placeholderMap.ContainsKey(placeHolder))
            {
                placeholderMap.Add(placeHolder, val);
            }
        }

        return placeholderMap.Aggregate(message, (current, pair) => current.Replace(pair.Key, pair.Value.ToString()));
    }

    private static string FormatWithAnonymousObjectOld(string message, object model)
    {
        var type = model.GetType();
        if (type == typeof(string))
        {
            return string.Format(message, model);
        }

        var placeHolders = PlaceHolderRegexOld().Matches(message).Select(m => m.Value).ToList();

        if (!placeHolders.Any())
        {
            return message;
        }

        var properties = type.GetProperties();
        var placeholderMap = new Dictionary<string, object>(placeHolders.Count);

        foreach (var placeHolder in placeHolders)
        {
            var propertyInfo = properties.FirstOrDefault(p => p.Name == placeHolder.Trim('{', '}'));

            // property found - extract value and add to the map
            var val = propertyInfo?.GetValue(model);
            if (val != null && !placeholderMap.ContainsKey(placeHolder))
            {
                placeholderMap.Add(placeHolder, val);
            }
        }

        return placeholderMap.Aggregate(message, (current, pair) => current.Replace(pair.Key, pair.Value.ToString()));
    }

    private static string FormatWithAnonymousObjectNew(string message, object model)
    {
        var type = model.GetType();
        if (type == typeof(string))
        {
            return string.Format(message, model);
        }

        var placeHolders = PlaceHolderRegexNew().Matches(message);

        if (placeHolders is not { Count: > 0 })
        {
            return message;
        }

        var properties = type.GetProperties();
        var placeholderMap = new Dictionary<string, object>(placeHolders.Count);

        foreach (Match placeHolderMatch in placeHolders)
        {
            if (placeHolderMatch is not { Success: true, Groups: { Count: > 1 } groups })
            {
                continue;
            }

            if (placeholderMap.ContainsKey(placeHolderMatch.Value))
            {
                // Don't process same placeholder twice
                continue;
            }

            var placeHolder = groups[1].Value;
            var propertyInfo = properties.FirstOrDefault(p => p.Name == placeHolder);

            // property found - extract value and add to the map, if it's not null
            var val = propertyInfo?.GetValue(model);
            if (val != null)
            {
                placeholderMap.Add(placeHolderMatch.Value, val);
            }
        }

        return placeholderMap.Aggregate(message, (current, pair) => current.Replace(pair.Key, pair.Value.ToString()));
    }

    [GeneratedRegex("{.+?}")]
    private static partial Regex PlaceHolderRegexOld();

    [GeneratedRegex("{(.+?)}")]
    private static partial Regex PlaceHolderRegexNew();
}
```

| Method                                           | Mean        | Error     | StdDev    | Gen0   | Gen1   | Allocated |
|------------------------------------------------- |------------:|----------:|----------:|-------:|-------:|----------:|
| FormatWithAnonymousObjectNoPlaceHolderOrig       |    64.53 ns |  1.004 ns |  0.939 ns | 0.0035 |      - |     176 B |
| FormatWithAnonymousObjectOrig                    |   494.81 ns |  2.669 ns |  2.084 ns | 0.0324 |      - |    1640 B |
| FormatWithAnonymousObjectRepeatedPlaceHolderOrig | 2,981.06 ns | 18.694 ns | 16.572 ns | 0.1907 |      - |    9632 B |
| FormatWithAnonymousObjectNoPlaceHolderOld        |    50.73 ns |  0.380 ns |  0.337 ns | 0.0035 |      - |     176 B |
| FormatWithAnonymousObjectOld                     |   392.54 ns |  1.419 ns |  1.258 ns | 0.0324 |      - |    1640 B |
| FormatWithAnonymousObjectRepeatedPlaceHolderOld  | 2,341.31 ns | 18.030 ns | 16.865 ns | 0.2022 |      - |   10192 B |
| FormatWithAnonymousObjectNoPlaceHolderNew        |    30.25 ns |  0.188 ns |  0.167 ns | 0.0017 |      - |      88 B |
| FormatWithAnonymousObjectNew                     |   431.95 ns |  2.986 ns |  2.647 ns | 0.0362 |      - |    1840 B |
| FormatWithAnonymousObjectRepeatedPlaceHolderNew  | 2,095.54 ns | 12.103 ns | 10.729 ns | 0.1717 | 0.0038 |    8792 B |

</details>

I'll take another look at this and try profiling the method to see if I accidentally introduced a new bottleneck

**Comment by stefanolsen** [comment]
@hangy Can you also replace `new CultureInfo(language)` with `CultureInfo.GetCultureInfo(language)` in `FallbackLanguagesCollection.cs`, since you are already working on that file?

**Comment by hangy** [comment]
> @hangy Can you also replace `new CultureInfo(language)` with `CultureInfo.GetCultureInfo(language)` in `FallbackLanguagesCollection.cs`, since you are already working on that file?

I can't find it there. You're talking about [this file](https://github.com/valdisiljuconoks/LocalizationProvider/blob/5be72f49c863e6cee723356e324f7cc63ed3bd2a/common/src/DbLocalizationProvider/FallbackLanguagesCollection.cs), correct?

**Comment by stefanolsen** [comment]
@hangy My bad. I was referring to [this](https://github.com/valdisiljuconoks/LocalizationProvider/blob/master/common/src/DbLocalizationProvider.Abstractions/LocalizationResourceTranslationCollection.cs).

**Comment by valdisiljuconoks** [comment]
@stefanolsen I'll finish `CultureInfo.Get(...)` in v9 branch

**Pull request merged**